src/SingleReport/MultiTouch.cpp

```diff
@@ -0,0 +1,246 @@
+	0xB1, 0x02,                    //   FEATURE (Data,Var,Abs)
+
+	0xC0                           // END_COLLECTION
+};
+
+
+MultiTouch_::MultiTouch_(void) : PluggableUSBModule(2, 1, epType), protocol(HID_REPORT_PROTOCOL), idle(1), featureReport((uint8_t *)&_ccmFeature), featureLength(sizeof(_ccmFeature))
+{
+	_report.reportID = _HID_REPORTID_TOUCH;
+	_ccmFeature.reportID = _HID_REPORTID_FEATURE;
+	_ccmFeature.contactCountMaximum = HID_MULTITOUCH_MAXFINGERS;
+	epType[0] = EP_TYPE_INTERRUPT_IN;
+	PluggableUSB().plug(this);
+}
+
+int MultiTouch_::getInterface(uint8_t* interfaceCount)
+{
+	*interfaceCount += 1; // uses 1
+	HIDDescriptor hidInterface = {
+		D_INTERFACE(pluggedInterface, 1, USB_DEVICE_CLASS_HUMAN_INTERFACE, HID_SUBCLASS_NONE, HID_PROTOCOL_NONE),
+		D_HIDREPORT(sizeof(_hidReportDescriptorTouchscreen_1) + sizeof(_hidReportDescriptorTouchscreen_2) * HID_MULTITOUCH_REPORTFINGERS + sizeof(_hidReportDescriptorTouchscreen_3)),
+		D_ENDPOINT(USB_ENDPOINT_IN(pluggedEndpoint), USB_ENDPOINT_TYPE_INTERRUPT, USB_EP_SIZE, 0x01)
+	};
+	return USB_SendControl(0, &hidInterface, sizeof(hidInterface));
+}
+
+int MultiTouch_::getDescriptor(USBSetup& setup)
+{
+	// Check if this is a HID Class Descriptor request
+	if (setup.bmRequestType != REQUEST_DEVICETOHOST_STANDARD_INTERFACE) { return 0; }
+	if (setup.wValueH != HID_REPORT_DESCRIPTOR_TYPE) { return 0; }
+
+	// In a HID Class Descriptor wIndex cointains the interface number
+	if (setup.wIndex != pluggedInterface) { return 0; }
+
+	// Reset the protocol on reenumeration. Normally the host should not assume the state of the protocol
+	// due to the USB specs, but Windows and Linux just assumes its in report mode.
+	protocol = HID_REPORT_PROTOCOL;
+
+	// Transmit HID descriptor. See comments next to the descriptor parts
+	int ret = 0;
+	ret += USB_SendControl(TRANSFER_PGM, _hidReportDescriptorTouchscreen_1, sizeof(_hidReportDescriptorTouchscreen_1));
+	for (int i = 0; i < HID_MULTITOUCH_REPORTFINGERS; i++) {
+		ret += USB_SendControl(TRANSFER_PGM, _hidReportDescriptorTouchscreen_2, sizeof(_hidReportDescriptorTouchscreen_2));
+	}
+	ret += USB_SendControl(TRANSFER_PGM, _hidReportDescriptorTouchscreen_3, sizeof(_hidReportDescriptorTouchscreen_3));
+
+	return ret;
+}
+
+bool MultiTouch_::setup(USBSetup& setup)
+{
+	if (pluggedInterface != setup.wIndex) {
+		return false;
+	}
+
+	uint8_t request = setup.bRequest;
+	uint8_t requestType = setup.bmRequestType;
+
+	if (requestType == REQUEST_DEVICETOHOST_CLASS_INTERFACE)
+	{
+		if (request == HID_GET_REPORT) {
+			if(setup.wValueH == HID_REPORT_TYPE_FEATURE){
+				// The only feature is Contact Count Maximum
+				USB_SendControl(0, featureReport, featureLength);
+				return true;
+			}
+			return true;
+		}
+		if (request == HID_GET_PROTOCOL) {
+			// TODO improve
+#ifdef __AVR__
+			UEDATX = protocol;
+#endif
+			return true;
+		}
+		if (request == HID_GET_IDLE) {
+			// TODO improve
+#ifdef __AVR__
+			UEDATX = idle;
+#endif
+			return true;
+		}
+	}
```

Comment:
Now I have a question on that:
My implementation is lacking support for the different modes. I guess this is also the reason why some keyboards do not work in bios mode. Can you please tell me more about that and why is this special AVR case added? Maybe we can improve the keyboard and mouse API as well?
https://github.com/NicoHood/HID/blob/master/src/SingleReport/BootMouse.cpp#L99-L109

Reply:
Those were copied and I didn't touch it. Sorry I have no idea why it's there either.

Reply:
Oh I see, those are from the keyboard API. I think that must be removed with care. I suggest implementing it as the mouse, linked above.

Reply:
Done